cpp/src/arrow/util/async_generator.h

```diff
@@ -1247,10 +1286,23 @@ class BackgroundGenerator {
           waiting_future.MarkFinished(next);
         }
       }
+      {
+        auto guard = state->mutex.Lock();
+        state->running_at_all = false;
+        if (state->finished) {
+          state->task_finished.MarkFinished();
+        }
+      }
     }
   };
 
   std::shared_ptr<State> state_;
+  // state_ is held by both the generator and the background thread so it won't be cleaned
+  // up when all consumer references are relinquished.  cleanup_ is only held by the
+  // generator so it will be destructed when the last consumer reference is gone.  We use
+  // this to cleanup / stop the background generator in case the consuming end stops
+  // listening (e.g. due to a downstream error)
+  std::shared_ptr<Cleanup> cleanup_;
```

Comment:
Would it be possible to simply use a `std::weak_ptr<State>` in the background thread?
(if there is some cleanup to do, you may do it in the State destructor?)

Reply:
Yes, I was torn on this myself.  If `weak_ptr` seems like a cleaner approach I can do that.

Reply:
Would be nice to give it a try and see what that gives.

Reply:
I started down this route and ran into a bit of difficulty.  The challenge is that I'd want to release the weak pointer while I do slow operations (otherwise I'm back at the original problem where the background thread keeps the state alive forever).  So right now the code look something like...

```
std::shared_ptr<State> current_state = state_.lock();
auto next = current_state->it.Next();
```

I could do something like...

```
std::shared_ptr<State> current_state = state_.lock();
auto& it = current_state->it;
current_state.reset();
auto next = it.Next();
```

...and it would actually be valid.  The state destructor would be waiting on task_finished before it returns so even though state might be in the middle of destruction it would not complete destruction until later.

In other words, when I realize I need to abandon I can't immediately delete state.  I have to signal the background worker to terminate immediately, wait for it to terminate, and then delete state.

So it works but I feel like the resulting code would be even harder to reason about because it would contain things that look invalid (accessing references obtained off a pointer after I give up the lifecycle tracking for that pointer) even though they just so happen to be valid.  That would complicate things for future readers I feel.

Reply:
I see, let's keep the cleanup object then. Thanks for the investigation!